crates/kitsune_p2p/types/src/tx2/tx2_utils/pool_buf.rs

```diff
@@ -0,0 +1,291 @@
+use std::cell::RefCell;
+
+// TODO - expirement with these values for efficiency.
+
+/// The max capacity of in-pool stored PoolBufs per thread.
+pub(crate) const POOL_MAX_CAPACITY: usize = 1024;
+
+/// Returned PoolBufs will be shrunk to this capacity when returned.
+pub(crate) const POOL_BUF_MAX_CAPACITY: usize = 4096;
+
+/// PoolBufs will be allocated/reset with this byte count BEFORE
+/// the readable buffer to make prepending frame info more efficient.
+pub(crate) const POOL_BUF_PRE_WRITE_SPACE: usize = 128;
+
+/// A buffer that will return to a pool after use.
+///
+/// When working with network code, we try to avoid two slow things:
+/// - Allocation
+/// - Initialization
+///
+/// We avoid allocation by returning used buffers to a pool for later re-use.
+///
+/// We avoid initialization by using `extend_from_slice()`.
+#[derive(Clone, PartialEq, Eq)]
+pub struct PoolBuf(Option<(usize, Vec<u8>)>);
```

Comment:
Seems good to me. Only feedback is we probably want to test this way more. I do have a feeling you could get the same behaviour with [bytes](https://docs.rs/bytes/1.0.1/bytes/index.html) inside a thread local. The advantage there would be it's already well tested but maybe it's missing functionality we need?

Reply:
I actually started using the bytes crate, especially give [tokio's read_buf](https://docs.rs/tokio/1.3.0/tokio/io/trait.AsyncReadExt.html#method.read_buf) - But the functionality I need isn't provided by bytes or vec, and all the functions I was using on bytes (such as extend_from_slice) just delegate directly to the Vec implementation, the same as I am doing. So, yeah, there was no benefit.

Agree we can write some more tests.